### PR TITLE
use jump table for expression execution

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Experimental PR to move expression evaluation dispatching from large switch to a jump.
Performance effects are unclear, but can be checked with this PR.

- [ ] :hankey: Bugfix
- [ ] :pizza: New feature
- [x] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.9: -
  - [ ] Backport for 3.8: -
  - [ ] Backport for 3.7: -

#### Related Information

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/Aql/Expression.cpp [hunk 1]
```diff
@@ -59,6 +59,123 @@ using namespace arangodb;
...
 using VelocyPackHelper = arangodb::basics::VelocyPackHelper;
 
+typedef AqlValue (*fptr)(ExpressionContext& ctx, AstNode const* node,
+                         bool& mustDestroy, bool doCopy);
+
+struct JumpTableInitializer {
+  JumpTableInitializer() {
+    for (size_t i = 0; i <= NODE_TYPE_WINDOW; ++i) {
+      table[i] = &InvalidHandler;
+    }
```
Comment (MBkkt) [inline arangod/Aql/Expression.cpp:69]: std::fill_n(table, NODE_TYPE_WINDOW + 1, &InvalidHandler)

File: arangod/Aql/Expression.cpp [hunk 2]
```diff
@@ -382,91 +499,6 @@ void Expression::prepareForExecution(ExpressionContext& ctx) {
   }
 }
 
-// brief execute an expression of type SIMPLE, the convention is that
-// the resulting AqlValue will be destroyed outside eventually
-AqlValue Expression::executeSimpleExpression(ExpressionContext& ctx,
-                                             AstNode const* node,
-                                             bool& mustDestroy, bool doCopy) {
-  switch (node->type) {
```
Comment (MBkkt) [inline arangod/Aql/Expression.cpp:390]: Btw can't compiler generate same code for this switch?
I thought switch everything better than manually written map.
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=78113
(all variant implementation using switch for visiting now (they use table when types count more than some number, [for msvc 256](https://github.com/microsoft/STL/blob/main/stl/inc/variant#L721)))

Comment (jsteemann) [reply arangod/Aql/Expression.cpp:390]: I was also disappointed that this change improved performance at least on the instances we tried it on.

**Comment by jsteemann** [comment]
Closing for now, because performance improvements are unclear and potentially not worth the complexity.

**Pull request closed without merging**